File: src/serpApi.ts
import SerpApiSearch from 'serpapi'

interface FlightSearchParams {
  origin: string
  destination: string
  date?: string
  preferences?: string
}

interface HotelSearchParams {
  location: string
  checkInDate?: string
  checkOutDate?: string
}

export async function searchFlights({ origin, destination, date, preferences }: FlightSearchParams): Promise<any> {
  const client = new SerpApiSearch(process.env.SERPAPI_KEY!)
  const params: Record<string, unknown> = {
    engine: 'google_flights',
    origin,
    destination
  }
  if (date) params.date = date
  if (preferences) params.preferences = preferences
  const result = await client.json(params)
  return result
}

export async function searchHotels({ location, checkInDate, checkOutDate }: HotelSearchParams): Promise<any> {
  const client = new SerpApiSearch(process.env.SERPAPI_KEY!)
  const params: Record<string, unknown> = {
    engine: 'google_hotels',
    location
  }
  if (checkInDate) params.checkin = checkInDate
  if (checkOutDate) params.checkout = checkOutDate
  const result = await client.json(params)
  return result
}

File: pyagent/serpapi.py
from serpapi import GoogleSearch
import os

async def search_flights(origin: str, destination: str, date: str = None, preferences: str = None):
    params = {
        'engine': 'google_flights',
        'api_key': os.environ['SERPAPI_KEY'],
        'origin': origin,
        'destination': destination
    }
    if date:
        params['date'] = date
    if preferences:
        params['preferences'] = preferences
    search = GoogleSearch(params)
    return search.get_json()

async def search_hotels(location: str, check_in: str = None, check_out: str = None):
    params = {
        'engine': 'google_hotels',
        'api_key': os.environ['SERPAPI_KEY'],
        'location': location
    }
    if check_in:
        params['checkin'] = check_in
    if check_out:
        params['checkout'] = check_out
    search = GoogleSearch(params)
    return search.get_json()
